File: json/src/main/java/com/ufoscout/coreutils/json/JsonSerializerService.java
package com.ufoscout.coreutils.json;

import java.io.OutputStream;

/**
 * 
 * @author Francesco Cina'
 *
 */
public interface JsonSerializerService {

	/**
	 * Return the json representation of the Bean
	 * @param object
	 * @return
	 */
	String toJson(Object object);

	/**
	 * Return the json representation of the Bean
	 * @param object
	 */
	void toJson(Object object, OutputStream out);
	
	/**
	 * Return the json representation of the Bean
	 * WARN: it is slower than the other method!
	 * @param object
	 * @return
	 */
	String toPrettyPrintedJson(Object object);

	/**
	 * Return the json representation of the Bean
	 * WARN: it is slower than the other method!
	 * @param object
	 */
	void toPrettyPrintedJson(Object object, OutputStream out);
	
	<T> T fromJson(Class<T> clazz, String json);

}

File: json-kotlin/src/main/java/com/ufoscout/coreutils/json/kotlin/JsonSerializerService.kt
package com.ufoscout.coreutils.json.kotlin

import com.ufoscout.coreutils.json.JsonSerializerService
import java.io.OutputStream
import kotlin.reflect.KClass

/**
 *
 * @author Francesco Cina'
 */
class JsonSerializerService(private val jsonService: JsonSerializerService) {

    /**
     * Return the json representation of the Bean
     * @param obj
     * @return
     */
    fun toJson(obj: Any): String {
        return jsonService.toJson(obj)
    }

    /**
     * Return the json representation of the Bean
     * @param obj
     */
    fun toJson(obj: Any, out: OutputStream) {
        jsonService.toJson(obj, out)
    }

    /**
     * Return the json representation of the Bean
     * WARN: it is slower than the other method!
     * @param obj
     * @return
     */
    fun toPrettyPrintedJson(obj: Any): String {
        return jsonService.toPrettyPrintedJson(obj)
    }

    /**
     * Return the json representation of the Bean
     * WARN: it is slower than the other method!
     * @param obj
     */
    fun toPrettyPrintedJson(obj: Any, out: OutputStream) {
        jsonService.toPrettyPrintedJson(obj, out)
    }

    /**
      * Method to deserialize JSON content from given JSON content String.
      */
    fun <T: Any> fromJson(clazz: KClass<T>, json: String): T {
        return jsonService.fromJson(clazz.java, json)
    }

    /**
     * Method to deserialize JSON content from given JSON content String.
     */
    inline fun <reified T: Any> fromJson(json: String) = fromJson(T::class, json)
}

